### PR TITLE
Add `unfinished-comments` feature

### LINKED ISSUE #3680: Notify the user of unfinished comments in a tab
How many times do you start writing a comment in a tab, the scroll up, change tab, and forget about it?

We could display this in the title/tab:

### PR DESCRIPTION
**Pull request by cheap-glitch**
Thanks for contributing! 🍄

1. LINKED ISSUES: Closes #3680 

2. TEST URLS:
- https://github.com/sindresorhus/refined-github/issues/3680
- https://github.com/sindresorhus/refined-github/pull/3596 (hidden textareas)

3. SCREENSHOT:
Peek 2020-10-31 18-24

Notes:
- I used the `visibilitychange` event instead of `pageshow`/`pagehide` as the latter weren't working for me in Firefox. As an added bonus, it also works when hiding/minifying the whole browser window.
- I followed the suggestion of the issue for the notification text, but imho something shorter would be better, e.g. `(+)` or `[*]` (akin to what editors show when a file has been modified but not saved).

### CODE REVIEW

File: source/features/unfinished-comments.tsx [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	if (document.body.dataset.visibilityListenerAdded) {
+		return;
+	}
+
+	const textareas = select.all('textarea');
+
+	document.addEventListener('visibilitychange', async () => {
+		if (document.visibilityState === 'hidden') {
+			if (textareas.some(textarea => (textarea.offsetWidth > 0 || textarea.offsetHeight > 0) && textarea.value.length > 0)) {
+				await cache.set(cacheKey, document.title);
```
Comment (fregante) [inline source/features/unfinished-comments.tsx:19]: A global is plenty, no need to literally store it in the cache 😃

File: source/features/unfinished-comments.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+const cacheKey = __filebasename + ':' + window.location.href;
+
+function init(): void {
+	if (document.body.dataset.visibilityListenerAdded) {
+		return;
+	}
+
+	const textareas = select.all('textarea');
+
+	document.addEventListener('visibilitychange', async () => {
```
Comment (fregante) [inline source/features/unfinished-comments.tsx:16]: This function needs to be moved to the top level so you don’t need to use dataset. There’s no problem with just selecting the textareas every time.

File: source/features/unfinished-comments.tsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+		if (document.visibilityState === 'hidden') {
+			if (textareas.some(textarea => (textarea.offsetWidth > 0 || textarea.offsetHeight > 0) && textarea.value.length > 0)) {
```
Comment (fregante) [inline source/features/unfinished-comments.tsx:18]: I’d use the length check first since it’s the lighter part

**Review by fregante — CHANGES_REQUESTED** [review]
Thank you!

**Comment by cheap-glitch** [comment]
What do you think about using a shorter text for the indicator (e.g. `[+]`)? It hides less of the original tab title — but on the other hand it's also less noticeable.

File: source/features/unfinished-comments.tsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+async function updateDocumentTitle(): Promise<void> {
+	if (document.visibilityState === 'hidden') {
+		if (select.all('textarea').some(textarea => textarea.value.length > 0 && (textarea.offsetWidth > 0 || textarea.offsetHeight > 0))) {
+			console.log('changing title');
+			documentTitle = document.title;
+			document.title = '(Draft comment) ' + document.title;
+		}
+	} else {
+		document.title = documentTitle;
```
Comment (fregante) [inline source/features/unfinished-comments.tsx:16]: We should only change the title if we changed it, and then we should reset it. The title might be changed by GitHub as well, so we don’t want to cache it for long.

File: source/features/unfinished-comments.tsx [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import * as pageDetect from 'github-url-detection';
+
+import features from '.';
+
+let documentTitle = '';
+
+async function updateDocumentTitle(): Promise<void> {
+	if (document.visibilityState === 'hidden') {
+		if (select.all('textarea').some(textarea => textarea.value.length > 0 && (textarea.offsetWidth > 0 || textarea.offsetHeight > 0))) {
+			console.log('changing title');
```
Comment (fregante) [inline source/features/unfinished-comments.tsx:11]: ?

**Comment by fregante** [comment]
I suggested the extended text because anything else doesn’t give enough information about what’s going on, especially since this isn’t a GitHub feature. Ideally you want to go back to the tab as soon as possible and finish the comment anyway.

File: source/features/unfinished-comments.tsx (hunk 4 above)
Comment (cheap-glitch) [reply source/features/unfinished-comments.tsx:15]: This should be fixed now, I think.

File: source/features/unfinished-comments.tsx [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+			document.title = '(Draft comment) ' + document.title;
+		}
+	} else if (documentTitle) {
+		document.title = documentTitle;
+		documentTitle = null;
+	}
+}
+
+function init(): void {
+	documentTitle = document.title;
```
Comment (fregante) [inline source/features/unfinished-comments.tsx:21]:
```suggestion
```

File: source/features/unfinished-comments.tsx [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import select from 'select-dom';
+import * as pageDetect from 'github-url-detection';
+
+import features from '.';
+
+let documentTitle: string | null = null;
+
+async function updateDocumentTitle(): Promise<void> {
+	if (document.visibilityState === 'hidden') {
```
Comment (fregante) [inline source/features/unfinished-comments.tsx:9]: If you flip this condition, you can avoid nesting the conditions:
```js
if (documentTitle) { // PR review-only comment: if this exists, we're 'visible'
	document.title = documentTitle;
	documentTitle = undefined; // PR review-only comment: never null https://github.com/sindresorhus/meta/issues/7
} else if (document.visibilityState === 'hidden' && hasDraftComments()) {
	documentTitle = document.title;
	document.title = '(Draft comment) ' + document.title;
}
```

File: source/features/unfinished-comments.tsx [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+		if (select.all('textarea').some(textarea => textarea.value.length > 0 && (textarea.offsetWidth > 0 || textarea.offsetHeight > 0))) {
```
Comment (fregante) [inline source/features/unfinished-comments.tsx:10]: We only ever need to check one dimension:
```suggestion
		if (select.all('textarea').some(textarea => textarea.value.length > 0 && textarea.offsetWidth > 0) {
```

**Comment by fregante** [comment]
It works!

**Review by fregante — APPROVED** [review]

**Comment by fregante** [comment]
Test

**Comment by fregante** [comment]
This unfortunately also catches comments that are being sent, for example:

1. Write comment
2. Click to send it
3. Immediately switch tab

When switching the tab, the comment will be detected as "draft" but will be sent not long after. I think that there should be an event we can listen to so that when the comment is submitted, the label is removed.

I wonder how this can be solved easily and comprehensively (`wait-for-ci` for example also will cause this)

**Comment by cheap-glitch** [comment]
> When switching the tab, the comment will be detected as "draft" but will be sent not long after. I think that there should be an event we can listen to so that when the comment is submitted, the label is removed.

We could try listening to the `submit` event of every form on the page.

**Pull request merged**